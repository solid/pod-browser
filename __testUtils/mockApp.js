/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createSolidDataset,
  createThing,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { vcard, rdf, schema } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { vcardExtras } from "../src/addressBook";

export const oidcPrefix = "http://www.w3.org/ns/solid/oidc#";
export const TOS_PREDICATE = `${oidcPrefix}tos_uri`;
export const POLICY_PREDICATE = `${oidcPrefix}policy_uri`;
export const CONTACTS_PREDICATE = `${oidcPrefix}contacts`;
export const LOGO_PREDICATE = `${oidcPrefix}logo_uri`;

export const APP_WEBID = "https://mockappurl.com";
export const APP_POLICY_URL = "https://mockappurl.com/privacy-policy";
export const APP_TOS_URL = "https://mockappurl.com/terms-of-service";

const contactThing = chain(
  createThing({ url: "https://mockappurl.com#contacts" }),
  (t) => addStringNoLocale(t, vcard.fn, "Example Contact"),
  (t) => addStringNoLocale(t, vcard.hasTelephone, "555-5555"),
  (t) => addStringNoLocale(t, vcard.hasEmail, "example@email.com"),
  (t) => addUrl(t, vcard.url, "https://examplewebsite.com")
);

const dataset = setThing(createSolidDataset(), contactThing);

export function mockApp() {
  return chain(
    createThing({ url: APP_WEBID }),
    (t) => addStringNoLocale(t, vcard.fn, "Mock App"),
    (t) => addUrl(t, vcardExtras("WebId"), APP_WEBID),
    (t) => addUrl(t, rdf.type, schema.SoftwareApplication), // just guessing what the type might be for apps
    (t) => addUrl(t, TOS_PREDICATE, APP_TOS_URL),
    (t) => addUrl(t, POLICY_PREDICATE, APP_POLICY_URL),
    (t) => addUrl(t, CONTACTS_PREDICATE, "https://mockappurl.com#contacts")
  );
}

export function mockAppDataset() {
  return setThing(dataset, mockApp());
}
