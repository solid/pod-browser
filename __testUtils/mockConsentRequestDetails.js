/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

async function getConsentRequestDetails() {
  return {
    "@context": [
      "https://www.w3.org/2018/credentials/v1",
      "https://w3id.org/security/suites/ed25519-2020/v1",
      "https://consent.pod.inrupt.com/credentials/v1",
    ],
    id: "https://consent.pod.inrupt.com/vc/53973727-f4d0-9e8dbdc041fd",
    type: ["VerifiableCredential", "SolidCredential", "SolidConsentRequest"],
    issuer: "https://consent.pod.inrupt.com",
    issuanceDate: "2021-05-26T16:40:03Z",
    expirationDate: "2021-06-23T16:40:03Z",
    credentialSubject: {
      id: "https://pod.inrupt.com/bob/profile/card#me",
      inbox: "https://pod.inrupt.com/bob/inbox/",
      hasConsent: {
        mode: ["Read", "Write"],
        hasStatus: "ConsentStatusRequested",
        forPersonalData: "https://pod.inrupt.com/alice/private/data",
        forPurpose: "https://example.com/SomeSpecificPurpose",
      },
    },
    proof: {
      created: "2021-05-26T16:40:03.009Z",
      proofPurpose: "assertionMethod",
      proofValue: "eqp8h_kL1DwJCpn65z-d1Arnysx6b11...jb8j0MxUCc1uDQ",
      type: "Ed25519Signature2020",
      verificationMethod: "https://consent.pod.inrupt.com/key/396f686b4ec",
    },
  };
}

export default getConsentRequestDetails;
