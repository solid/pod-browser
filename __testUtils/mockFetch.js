/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

export default function mockFetch(responses = {}) {
  return jest.fn().mockImplementation((url, init) => {
    const responseFn = responses[url.toString()];
    const responseObject = responseFn;
    const method = init?.method || "GET";
    let response;
    if (responseFn && typeof responseObject[method] !== "undefined") {
      response = responseObject[method](url, init);
    } else if (responseFn) {
      response = responseFn(url, init);
    } else {
      throw new Error(`URL (${url}) not mocked properly`);
    }
    Object.defineProperty(response, "url", { value: url });
    return Promise.resolve(response);
  });
}

export function mockSolidClientRequest(body, method = "GET", headers = {}) {
  return new Request(body, {
    headers: new Headers(headers),
    method,
  });
}

export function mockTurtleResponse(status, body = "") {
  return new Response(body, {
    status,
    headers: {
      "Content-Type": "text/turtle",
    },
  });
}
