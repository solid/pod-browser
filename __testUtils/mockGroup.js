/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  asUrl,
  getThing,
  mockSolidDatasetFrom,
  mockThingFrom,
  removeUrl,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { foaf, rdf, vcard } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { getBaseUrl } from "../src/solidClientHelpers/resource";

export function addMembersToMockedGroup(group, agentUrls) {
  const dataset = chain(
    group.dataset,
    ...agentUrls.map((agentUrl) => (d) =>
      setThing(d, setUrl(group.thing, vcard.hasMember, agentUrl))
    )
  );

  return {
    dataset,
    thing: getThing(dataset, asUrl(group.thing)),
  };
}

export function removeMembersFromMockedGroup(group, agentUrls) {
  const dataset = chain(
    group.dataset,
    ...agentUrls.map((agentUrl) => (d) =>
      setThing(d, removeUrl(group.thing, vcard.hasMember, agentUrl))
    )
  );

  return {
    dataset,
    thing: getThing(dataset, asUrl(group.thing)),
  };
}

export function mockGroupThing(name, url, { description, members } = {}) {
  return chain(
    mockThingFrom(url),
    (t) => setUrl(t, rdf.type, vcard.Group),
    (t) => setStringNoLocale(t, foaf.name, name),
    ...(members || []).map((agentUrl) => (t) =>
      addUrl(t, vcard.hasMember, agentUrl)
    ),
    ...[
      description ? (t) => setStringNoLocale(t, vcard.note, description) : null,
    ].filter((fn) => !!fn)
  );
}

export default function mockGroup(name, url, options = {}) {
  const thing = mockGroupThing(name, url, options);
  const dataset = setThing(mockSolidDatasetFrom(getBaseUrl(url)), thing);
  return { dataset, thing };
}
