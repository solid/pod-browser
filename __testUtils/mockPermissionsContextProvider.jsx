/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useState } from "react";
import T from "prop-types";
import PermissionsContext, {
  defaultPermissionsContext,
} from "../src/contexts/permissionsContext";

const permissions = [
  {
    acl: {
      read: true,
      write: true,
      append: false,
      control: false,
    },
    webId: "https://example1.com/profile/card#me",
    profile: {
      avatar: null,
      name: "Example 1",
      types: ["https://schema.org/Person"],
    },
    alias: "editors",
  },
  {
    acl: {
      read: true,
      write: true,
      append: false,
      control: false,
    },
    webId: "https://example2.com/profile/card#me",
    profile: {
      avatar: null,
      name: "Example 2",
      types: ["https://schema.org/Person"],
    },
    alias: "editors",
  },
  {
    acl: {
      read: true,
      write: true,
      append: false,
      control: false,
    },
    webId: "https://example3.com/profile/card#me",
    profile: {
      avatar: null,
      name: "Example 3",
      types: ["https://schema.org/Person"],
    },
    alias: "editors",
  },
  {
    acl: {
      read: true,
      write: true,
      append: false,
      control: false,
    },
    alias: "editors",
    webId: "https://example.com/profile/card#me",
    profile: {
      avatar: null,
      name: "Example 0",
      types: ["https://schema.org/Person"],
    },
  },
  {
    acl: {
      read: true,
      write: true,
      append: false,
      control: false,
    },
    alias: "editors",
    webId: "https://example4.com/profile/card#me",
    profile: {
      avatar: null,
      name: "Example 4",
      types: ["https://schema.org/Person"],
    },
    inherited: true,
  },
];

export default function mockBookmarksContextProvider(value) {
  function PermissionsContextProvider({ children }) {
    const [newAgentsWebIds, setNewAgentsWebIds] = useState(
      value?.newAgentsWebIds || []
    );
    const [webIdsToDelete, setWebIdsToDelete] = useState(
      value?.webIdsToDelete || []
    );
    const [addingWebId, setAddingWebId] = useState(value?.addingWebId || false);

    return (
      <PermissionsContext.Provider
        value={{
          ...defaultPermissionsContext,
          permissions: value?.permissions || permissions,
          webIdsToDelete,
          setWebIdsToDelete: value?.setWebIdsToDelete || setWebIdsToDelete,
          addingWebId,
          setAddingWebId: value?.setAddingWebId || setAddingWebId,
          newAgentsWebIds,
          setNewAgentsWebIds: value?.setNewAgentsWebIds || setNewAgentsWebIds,
          ...value,
        }}
      >
        {children}
      </PermissionsContext.Provider>
    );
  }

  PermissionsContextProvider.propTypes = {
    children: T.node,
  };

  PermissionsContextProvider.defaultProps = {
    children: null,
  };

  return PermissionsContextProvider;
}
