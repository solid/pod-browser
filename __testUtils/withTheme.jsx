/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { render } from "@testing-library/react";
import T from "prop-types";
import { StylesProvider, ThemeProvider } from "@inrupt/prism-react-components";
import { FeatureProvider } from "../src/contexts/featureFlagsContext";
import defaultTheme from "../src/theme";

export default function WithTheme(props) {
  const { theme, children } = props;
  return (
    <StylesProvider generateClassName={(rule) => `PodBrowser-${rule.key}`}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </StylesProvider>
  );
}

WithTheme.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  theme: T.object,
  children: T.node.isRequired,
};

WithTheme.defaultProps = {
  theme: defaultTheme,
};

export const renderWithTheme = (children, theme = defaultTheme) => {
  return render(
    <FeatureProvider>
      <WithTheme theme={theme}>{children}</WithTheme>
    </FeatureProvider>
  );
};
