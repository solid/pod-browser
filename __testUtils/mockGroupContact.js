/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  asUrl,
  getThing,
  mockSolidDatasetFrom,
  mockThingFrom,
  removeThing,
  removeUrl,
  setThing,
} from "@inrupt/solid-client";
import { chain } from "../src/solidClientHelpers/utils";
import { createGroupDatasetUrl } from "../src/models/contact/group";
import { vcardExtras } from "../src/addressBook";
import { mockGroupThing } from "./mockGroup";
import { getAddressBookDatasetUrl } from "../src/models/addressBook";

export function mockIndexThing(indexDataset, addressBook, groupThingUrl) {
  const addressBookThingUrl = asUrl(
    addressBook.thing,
    getAddressBookDatasetUrl(addressBook)
  );
  const addressBookThing =
    getThing(indexDataset, addressBookThingUrl) ||
    mockThingFrom(addressBookThingUrl);
  return addUrl(addressBookThing, vcardExtras("includesGroup"), groupThingUrl);
}

export function addGroupToMockedIndexDataset(
  dataset,
  addressBook,
  name,
  groupThingUrl
) {
  return chain(
    dataset,
    (d) => setThing(d, mockGroupThing(name, groupThingUrl)),
    (d) => setThing(d, mockIndexThing(d, addressBook, groupThingUrl))
  );
}

export function removeGroupFromMockedIndexDataset(
  dataset,
  addressBook,
  groupIndex,
  groupThing,
  groupThingUrl
) {
  return chain(
    dataset,
    (d) => removeThing(d, groupThing),
    (d) =>
      setThing(
        d,
        chain(
          getThing(groupIndex, asUrl(addressBook.thing)),
          (t) => removeUrl(t, vcardExtras("includesGroup"), groupThingUrl) // remove the includesGroup reference to group
        )
      )
  );
}

export default function mockGroupContact({
  addressBook,
  name,
  indexUrl = null,
  options = {},
}) {
  const { url, id } = options;
  const groupDatasetUrl = url || createGroupDatasetUrl(addressBook, id);
  const groupThingUrl = `${groupDatasetUrl}#this`;
  const groupThing = mockGroupThing(name, groupThingUrl, options);
  return {
    dataset: chain(
      mockSolidDatasetFrom(indexUrl || groupDatasetUrl),
      (d) => setThing(d, groupThing),
      (d) => setThing(d, mockIndexThing(d, addressBook, groupThingUrl))
    ),
    thing: groupThing,
  };
}
