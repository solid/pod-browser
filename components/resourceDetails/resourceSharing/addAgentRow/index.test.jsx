/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import userEvent from "@testing-library/user-event";
import { waitFor } from "@testing-library/dom";
import { useSession, useThing } from "@inrupt/solid-ui-react";
import { addUrl, createThing } from "@inrupt/solid-client";
import AddAgentRow, {
  EXISTING_WEBID_ERROR_MESSAGE,
  OWN_WEBID_ERROR_MESSAGE,
  TESTCAFE_ID_ADD_WEBID_BUTTON,
  TESTCAFE_ID_WEBID_INPUT,
} from "./index";
import { renderWithTheme } from "../../../../__testUtils/withTheme";
import mockPermissionsContextProvider from "../../../../__testUtils/mockPermissionsContextProvider";
import useContactProfile from "../../../../src/hooks/useContactProfile";
import { vcardExtras } from "../../../../src/addressBook";

jest.mock("@inrupt/solid-ui-react");
const mockedThingHook = useThing;

jest.mock("../../../../src/hooks/useContactProfile");
const mockedUseContactProfile = useContactProfile;
const permissions = [{ webId: "https://example.org/profile/card#me" }];
const setNewAgentsWebIds = jest.fn();
const PermissionsContextProvider = mockPermissionsContextProvider({
  permissions,
  addingWebId: true,
  setNewAgentsWebIds,
});

describe("AddAgentRow", () => {
  describe("when adding a new webId", () => {
    const mockThing = createThing();
    beforeEach(() => {
      mockedThingHook.mockReturnValue({ thing: mockThing });
      useSession.mockReturnValue({
        fetch: jest.fn(),
        session: { info: { webId: "https://podownerwebid.com/" } },
      });
      mockedUseContactProfile.mockReturnValue({ data: null });
    });
    const index = 0;
    const setNoAgentsAlert = jest.fn();
    const updateTemporaryRowThing = jest.fn();
    const contactsArrayLength = 0;
    it("renders a row with an input for the first item in the array", () => {
      const EmptyPermissionsContextProvider = mockPermissionsContextProvider({
        permissions: [],
        addingWebId: true,
      });
      const { asFragment } = renderWithTheme(
        <EmptyPermissionsContextProvider>
          <AddAgentRow
            contactsArrayLength={contactsArrayLength}
            type="editors"
            index={index}
            setNoAgentsAlert={setNoAgentsAlert}
            updateTemporaryRowThing={updateTemporaryRowThing}
          />
        </EmptyPermissionsContextProvider>
      );
      expect(asFragment()).toMatchSnapshot();
    });
    it("renders a validation error if added webId is already in permissions", () => {
      const { getByTestId, getByText } = renderWithTheme(
        <PermissionsContextProvider>
          <AddAgentRow
            contactsArrayLength={contactsArrayLength}
            type="editors"
            index={index}
            setNoAgentsAlert={setNoAgentsAlert}
            updateTemporaryRowThing={updateTemporaryRowThing}
          />
        </PermissionsContextProvider>
      );
      const input = getByTestId(TESTCAFE_ID_WEBID_INPUT);
      const addButton = getByTestId(TESTCAFE_ID_ADD_WEBID_BUTTON);
      userEvent.type(input, "https://example1.org/profile/card#me");
      userEvent.click(addButton);
      waitFor(() => {
        const errorText = getByText(EXISTING_WEBID_ERROR_MESSAGE);
        expect(errorText).not.toBeNull();
      });
    });
    it("renders a validation error if added webId is own webId", () => {
      const { getByTestId, getByText } = renderWithTheme(
        <PermissionsContextProvider>
          <AddAgentRow
            contactsArrayLength={contactsArrayLength}
            type="editors"
            index={index}
            setNoAgentsAlert={setNoAgentsAlert}
            updateTemporaryRowThing={updateTemporaryRowThing}
          />
        </PermissionsContextProvider>
      );
      const input = getByTestId(TESTCAFE_ID_WEBID_INPUT);
      const addButton = getByTestId(TESTCAFE_ID_ADD_WEBID_BUTTON);
      userEvent.type(input, "https://podownerwebid.com/");
      userEvent.click(addButton);
      const errorText = getByText(OWN_WEBID_ERROR_MESSAGE);
      expect(errorText).not.toBeNull();
    });
    it("clears error when changing input again", () => {
      const { getByTestId, queryByText } = renderWithTheme(
        <PermissionsContextProvider>
          <AddAgentRow
            contactsArrayLength={contactsArrayLength}
            type="editors"
            index={index}
            setNoAgentsAlert={setNoAgentsAlert}
            updateTemporaryRowThing={updateTemporaryRowThing}
            permissions={permissions}
          />
        </PermissionsContextProvider>
      );
      const input = getByTestId(TESTCAFE_ID_WEBID_INPUT);
      userEvent.type(input, "");
      const errorText = queryByText("That WebID has already been added");
      expect(errorText).toBeNull();
    });
    it("adds webId to webIds array on submit", () => {
      const { getByTestId } = renderWithTheme(
        <PermissionsContextProvider>
          <AddAgentRow
            contactsArrayLength={contactsArrayLength}
            type="editors"
            index={index}
            setNoAgentsAlert={setNoAgentsAlert}
            updateTemporaryRowThing={updateTemporaryRowThing}
            permissions={permissions}
          />
        </PermissionsContextProvider>
      );
      const input = getByTestId(TESTCAFE_ID_WEBID_INPUT);
      userEvent.type(input, "https://somewebid.com/");
      const addButton = getByTestId(TESTCAFE_ID_ADD_WEBID_BUTTON);
      userEvent.click(addButton);
      expect(setNewAgentsWebIds).toHaveBeenCalledWith([
        "https://somewebid.com/",
      ]);
    });
  });
  describe("with existing contacts", () => {
    const index = 0;
    const setNoAgentsAlert = jest.fn();
    const updateTemporaryRowThing = jest.fn();
    const contactsArrayLength = 1;

    beforeEach(() => {
      useSession.mockReturnValue({ fetch: jest.fn() });
    });

    it("renders a row with the agent's name if profile is available", () => {
      const mockThing = createThing();
      mockedThingHook.mockReturnValue({ thing: mockThing });
      mockedUseContactProfile.mockReturnValue({
        data: { webId: "https://somewebid.org", name: "Example", avatar: null },
      });
      const EmptyPermissionsContextProvider = mockPermissionsContextProvider({
        permissions: [],
        addingWebId: false,
      });
      const { asFragment, getByTestId, queryByText } = renderWithTheme(
        <EmptyPermissionsContextProvider>
          <AddAgentRow
            contactsArrayLength={contactsArrayLength}
            type="editors"
            index={index}
            setNoAgentsAlert={setNoAgentsAlert}
            updateTemporaryRowThing={updateTemporaryRowThing}
          />
        </EmptyPermissionsContextProvider>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(queryByText("Example")).toBeDefined();
      expect(getByTestId("agent-webid")).toBeDefined();
    });
    it("for newly added webids, if profile is unavailable it renders a row with the provided webId", () => {
      const mockThing = addUrl(
        createThing(),
        vcardExtras("WebId"),
        "https://somewebid.com"
      );
      mockedThingHook.mockReturnValue({ thing: mockThing });
      mockedUseContactProfile.mockReturnValue({
        data: null,
      });
      const EmptyPermissionsContextProvider = mockPermissionsContextProvider({
        permissions: [],
        addingWebId: false,
      });
      const { asFragment, getByTestId, queryByText } = renderWithTheme(
        <EmptyPermissionsContextProvider>
          <AddAgentRow
            contactsArrayLength={contactsArrayLength}
            type="editors"
            index={1}
            setNewAgentsWebIds={setNewAgentsWebIds}
            setNoAgentsAlert={setNoAgentsAlert}
            updateTemporaryRowThing={updateTemporaryRowThing}
          />
        </EmptyPermissionsContextProvider>
      );
      expect(asFragment()).toMatchSnapshot();
      expect(queryByText("https://somewebid.com")).toBeDefined();
      expect(getByTestId("agent-webid")).toBeDefined();
    });
  });
});
