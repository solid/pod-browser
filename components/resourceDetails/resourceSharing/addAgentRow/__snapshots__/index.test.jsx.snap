// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAgentRow when adding a new webId renders a row with an input for the first item in the array 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-agentPickerFormContainer"
  >
    <form
      class="PodBrowser-addAgentForm"
    >
      <div
        class="PodBrowser-searchBoxContainer"
      >
        <div
          aria-label="Enter WebID"
          class="PodBrowser-root PodBrowser-root PodBrowser-searchInput"
        >
          <div
            class="PodBrowser-root PodBrowser-root PodBrowser-searchInput PodBrowser-formControl PodBrowser-formControl"
            pattern="https://.+"
            title="Must start with https://"
          >
            <input
              aria-invalid="false"
              class="PodBrowser-input PodBrowser-input"
              data-testid="webid-input"
              placeholder="Enter WebID"
              required=""
              type="url"
              value=""
            />
          </div>
        </div>
      </div>
      <button
        class="PodBrowser-button PodBrowser-button--with-input PodBrowser-button"
        data-testid="add-button"
        type="submit"
      >
        <span
          class="PodBrowser-button__text"
        >
          Add
        </span>
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`AddAgentRow with existing contacts for newly added webids, if profile is unavailable it renders a row with the provided webId 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-nameAndAvatarContainer"
    title="Unable to load profile"
  >
    <div
      class="PodBrowser-root PodBrowser-circle PodBrowser-avatar PodBrowser-colorDefault"
    >
      <svg
        aria-hidden="true"
        class="PodBrowser-root PodBrowser-fallback"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
        />
      </svg>
    </div>
    <p
      class="PodBrowser-root PodBrowser-detailText PodBrowser-body1 PodBrowser-detailText"
      data-testid="agent-webid"
    />
  </div>
</DocumentFragment>
`;

exports[`AddAgentRow with existing contacts renders a row with the agent's name if profile is available 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-nameAndAvatarContainer"
    title="https://somewebid.org"
  >
    <div
      class="PodBrowser-root PodBrowser-circle PodBrowser-avatar PodBrowser-colorDefault"
    >
      <svg
        aria-hidden="true"
        class="PodBrowser-root PodBrowser-fallback"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
        />
      </svg>
    </div>
    <p
      class="PodBrowser-root PodBrowser-detailText PodBrowser-body1 PodBrowser-detailText"
      data-testid="agent-webid"
    >
      Example
    </p>
  </div>
</DocumentFragment>
`;
